pkg/registry/generic/etcd/etcd.go

```diff
@@ -45,7 +48,31 @@ type Etcd struct {
 	KeyRootFunc func(ctx api.Context) string
 
 	// Called for Create/Update/Get/Delete
-	KeyFunc func(ctx api.Context, id string) (string, error)
+	KeyFunc func(ctx api.Context, name string) (string, error)
+
+	// Called to get the name of an object
+	ObjectNameFunc func(obj runtime.Object) (string, error)
+
+	// Return the TTL objects should be persisted with. Update is true if this
+	// is an operation against an existing object.
+	TTLFunc func(obj runtime.Object, update bool) (uint64, error)
+
+	// Called on all objects returned from the underlying store, after
+	// the exit hooks are invoked.
+	Decorator rest.ObjectFunc
+	// Allows extended behavior during creation
+	CreateStrategy rest.RESTCreateStrategy
+	// On create of an object, attempt to run a further operation.
+	AfterCreate rest.ObjectFunc
+	// Allows extended behavior during updates
+	UpdateStrategy rest.RESTUpdateStrategy
+	// On update of an object, attempt to run a further operation.
+	AfterUpdate rest.ObjectFunc
+	// If true, return the object that was deleted. Otherwise, return a generic
+	// success status response.
+	ReturnDeletedObject bool
+	// On deletion of an object, attempt to run a further operation.
+	AfterDelete rest.ObjectFunc
 
 	// Used for all etcd access functions
 	Helper tools.EtcdHelper
```

Comment:
I think this is intentionally specific to etcd, specially TTLFunc.  Someone doing Zookeeper would copy generic.Etcd, and then copy pod/etcd/etcd.go.  The copied code would only be the stubs that expose the specific methods, which the test cases should validate.  Will add more godoc.

----- Original Message -----
> > +	// Called on all objects returned from the underlying store, after
> > +	// the exit hooks are invoked.
> > +	Decorator rest.ObjectFunc
> > +	// Allows extended behavior during creation
> > +	CreateStrategy rest.RESTCreateStrategy
> > +	// On create of an object, attempt to run a further operation.
> > +	AfterCreate rest.ObjectFunc
> > +	// Allows extended behavior during updates
> > +	UpdateStrategy rest.RESTUpdateStrategy
> > +	// On update of an object, attempt to run a further operation.
> > +	AfterUpdate rest.ObjectFunc
> > +	// If true, return the object that was deleted. Otherwise, return a
> > generic
> > +	// success status response.
> > +	ReturnDeletedObject bool
> > +	// On deletion of an object, attempt to run a further operation.
> > +	AfterDelete rest.ObjectFunc
> >  
> >  	// Used for all etcd access functions
> >  	Helper tools.EtcdHelper
> 
> This seems to be a lower-level storage abstraction than generic.Registry. If
> we were to port to another k-v store (say, Zookeeper, for Fabric8), we'd
> need something similar for that store. Is any of this interface really
> Etcd-specific? Please put a TODO in what you think is the right place to
> introduce that abstraction.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4419/files#r24682873